sonar-batch/src/main/java/org/sonar/batch/report/MeasuresPublisher.java

```diff
@@ -31,97 +33,135 @@
+      validateMeasure(input, resource.key());
+      return toReportMeasure(builder, input);
+    }
+
+    private static void validateMeasure(Measure measure, String componentKey) {
+      if (measure.getValue() == null && measure.getData() == null) {
+        throw new IllegalArgumentException(String.format("Measure on metric '%s' and component '%s' has no value, but it's not allowed", measure.getMetricKey(), componentKey));
+      }
+    }
+
+    private BatchReport.Measure toReportMeasure(BatchReport.Measure.Builder builder, Measure measure) {
+      builder.clear();
+
+      builder.setValueType(getMeasureValueType(measure.getMetric().getType()));
+      setValueAccordingToType(builder, measure);
+      // Because some numeric measures also have a data (like Sqale rating)
+      String data = measure.getData();
+      if (data != null) {
+        builder.setStringValue(data);
+      }
+      builder.setMetricKey(measure.getMetricKey());
+      return builder.build();
     }
-  }
 
-  private static void validateMeasure(Measure measure, String componentKey) {
-    if (measure.getValue() == null && measure.getData() == null) {
-      throw new IllegalArgumentException(String.format("Measure on metric '%s' and component '%s' has no value, but it's not allowed", measure.getMetricKey(), componentKey));
+    private void setValueAccordingToType(BatchReport.Measure.Builder builder, Measure measure) {
+      Serializable value = measure.value();
+      switch (builder.getValueType()) {
+        case BOOLEAN:
+          builder.setBooleanValue((Boolean) value);
+          break;
+        case DOUBLE:
+          builder.setDoubleValue(((Number) value).doubleValue());
+          break;
+        case INT:
+          builder.setIntValue(((Number) value).intValue());
+          break;
+        case LONG:
+          builder.setLongValue(((Number) value).longValue());
+          break;
+        case STRING:
+          builder.setStringValue((String) value);
+          break;
+        default:
+          throw new IllegalStateException("Unknown value type: " + builder.getValueType());
+      }
     }
+
+    private MeasureValueType getMeasureValueType(ValueType type) {
+      switch (type) {
+        case INT:
+        case RATING:
+          return MeasureValueType.INT;
+        case FLOAT:
+        case PERCENT:
+          return MeasureValueType.DOUBLE;
+        case BOOL:
+          return MeasureValueType.BOOLEAN;
+        case STRING:
+        case DATA:
+        case LEVEL:
+        case DISTRIB:
+          return MeasureValueType.STRING;
+        case WORK_DUR:
+        case MILLISEC:
+          return MeasureValueType.LONG;
+        default:
+          throw new IllegalStateException("Unknown value type: " + type);
+      }
+    }
+
   }
 
-  private BatchReport.Measure toReportMeasure(BatchReport.Measure.Builder builder, Measure measure) {
-    builder.clear();
+  private final class IsMetricAllowed implements Predicate<Measure> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)